File: mch1cpp/src/signal_processing_cpp/src/process.cpp
// Include libraries
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32.hpp"
#include <cmath>

// Class to process the sine signal
class SignalProcessor : public rclcpp::Node {
public:
    SignalProcessor() : Node("process"), time_(0.0), phase_shift_(M_PI / 4), amplitude_factor_(0.5), offset_(1.0) { // Select time shift factor, make the wave positive and redice its amplitude
        signal_subscription_ = this->create_subscription<std_msgs::msg::Float32>(
            "/signal", 10, std::bind(&SignalProcessor::process_signal, this, std::placeholders::_1));
        
        time_subscription_ = this->create_subscription<std_msgs::msg::Float32>(
            "/time", 10, std::bind(&SignalProcessor::store_time, this, std::placeholders::_1));

        processed_signal_publisher_ = this->create_publisher<std_msgs::msg::Float32>("/proc_signal", 10);
    }

private:
    void store_time(const std_msgs::msg::Float32::SharedPtr msg) {
        time_ = msg->data;
    }

    void process_signal(const std_msgs::msg::Float32::SharedPtr msg) {
        double original_signal = msg->data;
        // Process the signal with the desired transformations
        double shifted_signal = std::sin(time_ + phase_shift_);
        double processed_signal = amplitude_factor_ * (shifted_signal + offset_);
        std_msgs::msg::Float32 processed_msg;
        processed_msg.data = processed_signal;
        processed_signal_publisher_->publish(processed_msg); // Publish the new sine wave value to the topic
        RCLCPP_INFO(this->get_logger(), "Processed Signal: %.2f", processed_signal); // Print the processed signal value in the terminal
    }

    rclcpp::Subscription<std_msgs::msg::Float32>::SharedPtr signal_subscription_;
    rclcpp::Subscription<std_msgs::msg::Float32>::SharedPtr time_subscription_;
    rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr processed_signal_publisher_;

    double time_;
    const double phase_shift_;
    const double amplitude_factor_;
    const double offset_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SignalProcessor>());
    rclcpp::shutdown();
    return 0;
}

File: MCH1_WS/src/signal_processing/signal_processing/process.py
# Import libraries 
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32
import numpy as np

# Processing the sine wave
class SignalProcessor(Node):
    def __init__(self):
        super().__init__('process')
        self.subscription_signal = self.create_subscription(Float32, '/signal', self.process_signal, 10)
        self.subscription_time = self.create_subscription(Float32, '/time', self.store_time, 10)
        self.processed_signal_publisher = self.create_publisher(Float32, '/proc_signal', 10)
        self.phase_shift = np.pi / 4  # Phase shift
        self.offset = 1.0  # Make the sine wave to always be positive
        self.amplitude_factor = 0.5  # Reduce amplitude by half
        self.time = 0.0

    def store_time(self, msg):
        self.time = msg.data

    def process_signal(self, msg):
        original_signal = msg.data

        # Apply phase shift, offset, and amplitude reduction
        shifted_signal = np.sin(self.time + self.phase_shift)  
        processed_signal = self.amplitude_factor * (shifted_signal + self.offset)  

        # Publish the processed signal to the topic
        self.processed_signal_publisher.publish(Float32(data=processed_signal))
        self.get_logger().info(f'Processed Signal: {processed_signal:.2f}')

def main(args=None):
    rclpy.init(args=args)
    node = SignalProcessor()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
